### PR TITLE
Preserve ranges in indexing with IdentityUnitRange(::Base.OneTo)

### PR DESCRIPTION
**Pull request by jishnub**
This PR changes the type in indexing `AbstractRange`s with `IdentityUnitRange(::Base.OneTo)`. This is technically breaking so I've separated this from #210 . This may go into a minor release.

The reason this PR is necessary is that currently `OffsetArrays` pirates the `Base` behavior and changes the return type (although not the values/axes). 

After #210:
```julia

julia> r = 1:2:10
1:2:9

julia> s = Base.IdentityUnitRange(axes(r,1));

julia> r[s]
1:2:9

julia> r[s] |> typeof
StepRange{Int64,Int64}

julia> using OffsetArrays

julia> r[s]
1:2:9 with indices 1:5

julia> r[s] |> typeof
OffsetArray{Int64,1,StepRange{Int64,Int64}}
```

After this PR:
```julia
julia> using OffsetArrays

julia> r[s]
1:2:9

julia> r[s] |> typeof
StepRange{Int64,Int64}
```

Despite the piracy, the indexing operation returns the same type as `Base`. Importantly this returns an `AbstractRange` as opposed to an `OffsetVector`.

### CODE REVIEW

**Comment by timholy** [comment]
Just to clarify, this is breaking only in the sense that different types are returned, right? As you know, I'm generally in favor of thinking of types as an implementation detail so that kind of "breakage" is OK if it doesn't really change behavior.

**Comment by jishnub** [comment]
Yes this only changes the type of the returned object, there is no change in the axes or the values. While this is not breaking in that sense, it might cause some overly-specialized functions to falter. Perhaps it's better to be conservative and make this a part of 1.7 while #210 can go into 1.6.x?

In general it seems like a good strategy to push changes in types to minor releases. This way a compat entry that restricts the minor version would receive (mostly) non-breaking patches. There might be some permissible breakage in minor releases due to changes that are not considered a part of the API, whereas API-breaking changes go to major releases. This avoids patch releases breaking something as in #163

**Comment by jishnub** [comment]
The latest change in da226f7 is another step in preserving `AbstractUnitRanges` in vector indexing operations involving `OffsetUnitRanges`, as these may be used further in indexing, and indexing is more efficient with `AbstractUnitRanges` than with `OffsetUnitRanges`.

As a consequence, before:
```julia
julia> a = ones(200);

julia> r = OffsetVector(2:500, 2);

julia> s = OffsetVector(3:150, 3);

julia> @btime $a[$r[$s]];
  639.861 ns (1 allocation: 1.33 KiB)
```

Now:
```julia
julia> @btime $a[$r[$s]];
  356.967 ns (1 allocation: 1.33 KiB)
```

**Comment by jishnub** [comment]
Changes introduced in c3ef299: `to_indices` specialized for `OffsetUnitRange{<:Integers}`s by replacing them with `IdOffsetRange`s in the indexing operation. 

Previously:

```julia
julia> a = ones(200);

julia> s  = OffsetVector(3:150, 3);

julia> @btime $a[$(UnitRange(parent(s)))];
  188.573 ns (1 allocation: 1.33 KiB)

julia> @btime $a[$s];
  638.503 ns (1 allocation: 1.33 KiB)
```

After this

```julia
julia> @btime $a[$s];
  293.536 ns (1 allocation: 1.33 KiB)
```

The indexing performance becomes comparable to that of the parent

**Comment by jishnub** [comment]
We're back to 

```julia
julia> @btime $a[$r[$s]];
  358.557 ns (1 allocation: 1.33 KiB)

julia> @btime $a[$(r[s])];
  346.858 ns (1 allocation: 1.33 KiB)
```

File: benchmark/benchmarks.jl [hunk 1]
```diff
@@ -20,24 +23,37 @@ unsafe_update(x) = @inbounds(for i in axes(x,1); x[i] = x[i] + i; end)
...
-@show @benchmark unsafe_update(x)
-@show @benchmark unsafe_update(y)
-@show @benchmark unsafe_update2d(x2d)
-@show @benchmark unsafe_update2d(y2d)
-@show @benchmark unsafe_update_eachindex(x)
-@show @benchmark unsafe_update_eachindex(y)
+vectorindexing(a, s) = a[s]
+nestedvectorindexing(a, r, s) = a[r[s]]
+
+macro showbenchmark(ex)
```
Comment (jishnub) [inline benchmark/benchmarks.jl:29]: This is added to avoid cluttering the display with line numbers

File: src/OffsetArrays.jl [hunk 2]
```diff
@@ -333,35 +333,59 @@ Broadcast.broadcast_unalias(dest::OffsetArray, src::OffsetArray) = parent(dest)
...
+        end
     end
 
     # this method is needed for ambiguity resolution
-    @eval @propagate_inbounds Base.getindex(r::StepRangeLen{T,<:Base.TwicePrecision,<:Base.TwicePrecision}, s::$OR) where T =
-    OffsetArray(r[UnitRange(s)], axes(s))
-
-    #= Integer UnitRanges may return an appropriate AbstractUnitRange{<:Integer}, as the result may be used in indexing, and
-    indexing is faster with ranges =#
-    @eval @propagate_inbounds function Base.getindex(r::UnitRange{<:Integer}, s::$OR)
```
Comment (jishnub) [inline src/OffsetArrays.jl:361]: This is not necessary anymore as `_maybewrapoffset` already handles this case

File: src/utils.jl [hunk 3]
```diff
@@ -72,10 +72,10 @@ function _checkindices(N::Integer, indices, label)
     N == length(indices) || throw_argumenterror(N, indices, label)
 end
 
-_maybewrapaxes(A::AbstractVector, ::Base.OneTo) = no_offset_view(A)
```
Comment (jishnub) [inline src/utils.jl:75]: `_maybewrapaxes` is effectively equivalent to `_maybewrapoffset`, so there's no need for this

File: test/runtests.jl [hunk 4]
```diff
@@ -887,20 +910,32 @@ end
...
 @testset "Vector indexing with offset ranges" begin
     r = OffsetArray(8:10, -1:1)
     r1 = r[0:1]
-    @test r1 === 9:10
+    @test r1 == 9:10
     r1 = (8:10)[OffsetArray(1:2, -5:-4)]
-    @test axes(r1) == (IdentityUnitRange(-5:-4),)
-    @test parent(r1) === 8:9
+    @test axes(r1) == (-5:-4,)
+    @test no_offset_view(r1) == 8:9
```
Comment (jishnub) [inline test/runtests.jl:935]: These indexing operation returns an `IdOffsetRange` now, so the parent might not be equal to the expected range. We check the values instead using `no_offset_view` and the axes separately

**Comment by jishnub** [comment]
There is still some performance hit in Cartesian indexing:

```julia
julia> a = ones(2000, 2000);

julia> s = OffsetVector(axes(a,1), 0);

julia> @btime $a[$(parent(s)), $(parent(s))]; # no offset, so this is fast
  7.548 ms (2 allocations: 30.52 MiB)

julia> @btime $a[$(parent(s)), $s]; # not sure why this is faster?
  7.586 ms (2 allocations: 30.52 MiB)

julia> @btime $a[$s, $(parent(s))];
  7.841 ms (2 allocations: 30.52 MiB)

julia> @btime $a[$s, $s];
  7.875 ms (5 allocations: 30.52 MiB)
```

**Comment by jishnub** [comment]
Inlining `to_indices` got rid of the Cartesian indexing performance hit.

Now 
```julia
julia> @btime $a[$(parent(s)), $(parent(s))];
  6.755 ms (2 allocations: 30.52 MiB)

julia> @btime $a[$s, $s];
  6.718 ms (5 allocations: 30.52 MiB)
```

However, oddly it seems nested linear indexing is faster if `::OffsetUnitRange[::OffsetUnitRange]` returns an `OffsetRange` rather than an `IdOffsetRange`. I have reverted the indexing behavior to return `OffsetRange`s for now. 

```julia
julia> a = ones(2000);

julia> r = OffsetVector(1:2000, 0);

julia> s = OffsetVector(axes(a,1), 0);

julia> @btime $a[$r[$s]]; # r[s] Returns an OffsetRange
  3.465 μs (1 allocation: 15.75 KiB)

julia> @btime $a[$r[$s]]; # r[s] Returns an IdOffsetRange
  4.331 μs (1 allocation: 15.75 KiB)
```

I'm not sure what causes this hit, but in any case `r[s]` may return an `OffsetRange`, now that `to_indices` does the conversion to an `IdOffsetRange` internally. Will have to check some of the other return types and their performance impacts.

**Comment by jishnub** [comment]
I think we're almost there after 72a5eaf. Now

```julia
julia> a = ones(2000);

julia> r = OffsetVector(1:2000, 0);

julia> s = OffsetVector(axes(a,1), 0);

julia> @btime $a[$r[$s]];
  2.103 μs (1 allocation: 15.75 KiB)

julia> @btime $a[$s];
  2.067 μs (1 allocation: 15.75 KiB)
```

**Comment by jishnub** [comment]
Worth waiting for https://github.com/JuliaLang/julia/pull/39896 to restrict some of these methods to older versions of Julia

File: src/OffsetArrays.jl [hunk 5]
```diff
@@ -335,36 +335,83 @@ Broadcast.broadcast_unalias(dest::OffsetArray, src::OffsetArray) = parent(dest)
...
-const OffsetUnitRange{T} = OffsetArray{T,1,<:AbstractUnitRange{T}}
+const OffsetRange{T} = OffsetVector{T,<:AbstractRange{T}}
+const OffsetUnitRange{T} = OffsetVector{T,<:AbstractUnitRange{T}}
 const IIUR = IdentityUnitRange{S} where S<:AbstractUnitRange{T} where T<:Integer
 
 Base.step(a::OffsetRange) = step(parent(a))
 
-@propagate_inbounds function Base.getindex(a::OffsetRange, r::OffsetRange)
-    OffsetArray(a.parent[r.parent .- a.offsets[1]], axes(r))
+Base.checkindex(::Type{Bool}, inds::AbstractUnitRange, or::OffsetRange) = Base.checkindex(Bool, inds, parent(or))
```
Comment (timholy) [inline src/OffsetArrays.jl:344]: This doesn't seem correct:
```julia
julia> ax = OffsetArrays.IdOffsetRange(Base.OneTo(5), -3)
OffsetArrays.IdOffsetRange(values=-2:2, indices=-2:2)

julia> axes(ax) == axes(ax.parent)
false
```

Comment (jishnub) [reply src/OffsetArrays.jl:344]: Not sure I understand. `checkindex` does not use the axes, it checks that the first and last values are contained in `inds`, and the first and last values of an `OffsetRange` should be the same as those of its parent.

**Review by timholy — APPROVED** [review]
Sorry for the delay, LGTM. I like limiting the number of methods where we can, so making it version-dependent seems like a great idea.

**Pull request merged**